library/ccm.c

```diff
@@ -263,39 +246,51 @@ int mbedtls_ccm_set_lengths( mbedtls_ccm_context *ctx,
     if( total_ad_len >= 0xFF00 )
         return( MBEDTLS_ERR_CCM_BAD_INPUT );
 
-    if( ctx->state & CCM_STATE__ERROR )
-    {
-        mbedtls_ccm_clear_state(ctx);
-    }
-
     ctx->plaintext_len = plaintext_len;
     ctx->add_len = total_ad_len;
     ctx->tag_len = tag_len;
     ctx->processed = 0;
 
     ctx->state |= CCM_STATE__LENGHTS_SET;
-    return mbedtls_ccm_calculate_first_block(ctx);
+    return ccm_calculate_first_block_if_ready(ctx);
 }
 
 int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
                            const unsigned char *add,
                            size_t add_len )
 {
-    CCM_VALIDATE_RET( ctx->add_len == 0 || add != NULL );
-
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     unsigned char i;
     size_t olen, use_len, offset;
 
-    if( ctx->add_len > 0 && add_len > 0)
+    if( ctx->state & CCM_STATE__ERROR )
     {
+        return ret;
+    }
+
+    if( ctx->add_len > 0 && add_len > 0 )
+    {
+        if( ctx->state & CCM_STATE__AUTH_DATA_FINISHED )
+        {
+            return ret;
```

Comment:
> why error out on extra additional data only when ctx->add_len > 0?

Actually the `ctx->add_len > 0` is not really related to the `ctx->state & CCM_STATE__AUTH_DATA_FINISHED`.
The `ctx->add_len > 0` is there to ignore any input if configured auth data length was zero.

As for the error in case of `ctx->state & CCM_STATE__AUTH_DATA_FINISHED` (auth data exceeded set length), it is dependant on the `add_len` rather then `ctx->add_len`. We do not want the function to return an error in following scenario:
1. Pass all auth data to ccm_update_ad
2. Call ccm_update_ad with empty input (zero length)